ipc-queue/src/fifo.rs

```diff
@@ -0,0 +1,404 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use super::*;
+use fortanix_sgx_abi::FifoDescriptor;
+use std::cell::UnsafeCell;
+use std::mem;
+use std::ptr::NonNull;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+pub fn bounded<T, S>(len: usize, s: S) -> (Sender<T, S>, Receiver<T, S>)
+where
+    T: Identified,
+    S: Synchronizer,
+{
+    let arc = Arc::new(Fifo::new(len));
+    let inner = FifoInner::from_arc(arc);
+    let tx = Sender { inner: inner.clone(), synchronizer: s.clone() };
+    let rx = Receiver { inner, synchronizer: s };
+    (tx, rx)
+}
+
+pub fn bounded_async<T, S>(len: usize, s: S) -> (AsyncSender<T, S>, AsyncReceiver<T, S>)
+where
+    T: Identified,
+    S: AsyncSynchronizer,
+{
+    let arc = Arc::new(Fifo::new(len));
+    let inner = FifoInner::from_arc(arc);
+    let tx = AsyncSender { inner: inner.clone(), synchronizer: s.clone() };
+    let rx = AsyncReceiver { inner, synchronizer: s };
+    (tx, rx)
+}
+
+struct Fifo<T> {
+    data: Box<[T]>,
+    offsets: Box<AtomicUsize>,
+}
+
+impl<T: Identified> Fifo<T> {
+    fn new(len: usize) -> Self {
+        assert!(
+            len.is_power_of_two(),
+            "Fifo len should be a power of two"
+        );
+        let mut data = Vec::with_capacity(len);
+        data.resize_with(len, T::empty);
+        Self {
+            data: data.into_boxed_slice(),
+            offsets: Box::new(AtomicUsize::new(0)),
+        }
+    }
+}
+
+enum Storage<T> {
+    Shared(Arc<Fifo<T>>),
+    Static,
+}
+
+impl<T> Clone for Storage<T> {
+    fn clone(&self) -> Self {
+        match self {
+            Storage::Shared(arc) => Storage::Shared(arc.clone()),
+            Storage::Static => Storage::Static,
+        }
+    }
+}
+
+pub(crate) struct FifoInner<T> {
+    data: NonNull<UnsafeCell<[T]>>,
```

Comment:
Maybe an `[UnsafeCell<T>]` makes more sense. What do you think?

Reply:
Not sure if that makes any difference. Do you think `UnsafeCell<[T]>` could imply anything beyond the interior mutability of slice elements?

Reply:
With `[UnsafeCell<T>]` you can index/get the length without going through the unsafe wrapper. With `UnsafeCell<[T]>` you can (unsafely) change the pointer/length. Both of these points make me think `[UnsafeCell<T>]` is better.